dkg/src/dkg_manager/mod.rs

```diff
@@ -40,29 +95,199 @@ impl<DKG: DKGTrait, P: DKGPrivateParamsProvider<DKG>> DKGManager<DKG, P> {
+            let handling_result = tokio::select! {
+                dkg_start_event = dkg_start_event_rx.select_next_some() => {
+                    self.process_dkg_start_event(dkg_start_event).await
+                },
+                (_sender, msg) = rpc_msg_rx.select_next_some() => {
+                    self.process_peer_rpc_msg(msg).await
+                },
+                agg_node = agg_trx_rx.select_next_some() => {
+                    self.process_aggregated_transcript(agg_node).await
+                },
+                dkg_txn = dkg_txn_pulled_rx.select_next_some() => {
+                    self.process_dkg_txn_pulled_notification(dkg_txn).await
+                },
                 close_req = close_rx.select_next_some() => {
-                    self.vtxn_pool_write_cli.put(None);
-                    if let Ok(ack_sender) = close_req {
-                        ack_sender.send(()).unwrap();
-                    }
-                    break;
+                    self.process_close_cmd(close_req.unwrap())
                 }
+            };
+
+            match handling_result {
+                Err(e) => error!("{}", e),
+                _ => {},
             }
         }
     }
+
+    /// On a CLOSE command from epoch manager, do clean-up.
+    fn process_close_cmd(&mut self, ack_tx: oneshot::Sender<()>) -> Result<()> {
+        self.stopped = true;
+
+        if let InnerState::InProgress { abort_handle, .. } = &self.state {
+            abort_handle.abort();
+        }
+
+        self.vtxn_pool_write_cli.put(None);
+
+        let _ = ack_tx.send(());
+        Ok(())
+    }
+
+    /// On a vtxn pull notification, record metric.
+    async fn process_dkg_txn_pulled_notification(
+        &mut self,
+        _txn: Arc<ValidatorTransaction>,
+    ) -> Result<()> {
+        if let InnerState::Finished {
+            start_time_us,
+            pull_confirmed,
+            ..
+        } = &mut self.state
+        {
+            if !*pull_confirmed {
+                // TODO(zjma): metric DKG_AGG_NODE_PROPOSED
+            }
+            *pull_confirmed = true;
+        }
+        Ok(())
+    }
+
+    /// Calculate DKG config. Deal a transcript. Start broadcasting the transcript.
+    /// Called when a DKG start event is received, or when the node is restarting.
+    ///
+    /// NOTE: the dealt DKG transcript does not have to be persisted:
+    /// it is ok for a validator to equivocate on its DKG transcript, as long as the transcript is valid.
+    async fn setup_deal_broadcast(
+        &mut self,
+        start_time_us: u64,
+        target_validator_set: &ValidatorSet,
+    ) -> Result<()> {
+        self.state = match &self.state {
+            InnerState::NotStarted => {
+                let public_params = DKG::new_public_params(
+                    self.epoch_state.as_ref(),
+                    self.my_addr,
+                    target_validator_set,
+                );
+                let mut rng = thread_rng();
```

Comment:
I think I asked the question before but does this need to be persisted? or different values matters? e.g. crash recovery produces different transcript?

Reply:
now nothing needs to be persisted. Any trx is fine as long as it's verifiable.

Reply:
worth adding a comment here

Reply:
yes, we have some comments not far, on top of `setup_deal_broadcast`